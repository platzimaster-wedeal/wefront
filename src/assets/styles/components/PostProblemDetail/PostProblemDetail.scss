@import '../../partials/sizes.scss';
@import '../../partials/colors.scss';
@import '../../Breakpoints.scss';

.post-problem-detail {
  box-sizing: border-box;
  position: relative;
  width: 100%;
  height: auto;
  display: flex;
  flex-flow: column nowrap;
  align-items: flex-start;
  padding: $space-size-big;

  &__actions {
    display: flex;
    flex-flow: row nowrap;
    width: auto;
    height: auto;
    padding: $space-size-normal 0;
  }

}
