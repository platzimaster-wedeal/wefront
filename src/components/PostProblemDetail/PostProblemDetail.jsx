import React, { useEffect, useState } from "react";

// Hooks
import { useInputForm } from "../../hooks/useInputForm/useInputForm";
import { useModal } from "../../hooks/useModal/useModal";

// Redux
import { useDispatch, useSelector } from "react-redux";
import { APPLY_PROBLEM } from "../../redux/types/Problems/ProblemsTypes";

// Services
import { postScoreUser } from "../../services/UserService/userService";
import {
 getJobOffer,
 getWorkersJob,
 getFullProblem,
} from "../../services/ProblemsService/problemsService";

// Styles
import "../../assets/styles/components/PostProblemDetail/PostProblemDetail.scss";

// Components
import PostHeader from "../PostHeader/PostHeader";
import PostProblemDescription from "../PostProblemDescription/PostProblemDescription";
import PostProblemSchedule from "../PostProblemSchedule/PostProblemSchedule";
import PostProblemPayment from "../PostProblemPayment/PostProblemPayment";
import PostProblemPrevisualization from "../PostProblemPrevisualization/PostProblemPrevisualization";
import PostProblemRequirements from "../PostProblemRequirements/PostProblemRequirements";
import SelectDate from "../DatePicker/SelectDate";
import Button from "../Buttons/Button";
import PostProblemStatus from "../PostProblemStatus/PostProblemStatus";
import ProblemStatusInformation from "../ProblemStatusInformation/ProblemStatusInformation";
import CreateApplyPostProblem from "../CreateApplyPostProblem/CreateApplyPostProblem";
import Loading from "../Loading/Loading";

// Modal
import ModalContainer from "../Modals/ModalContainer";
import Modal from "../Modals/Modal";
import ModalMessage from "../ModalMessage/ModalMessage";

const PostProblemDetail = ({
 problem,
 idJobOffer,
 setDataApply,
 idUser,
 idEmployee,
 isError = null,
 isLoading = true,
 onApply,
}) => {
 const { id } = useSelector((state) => state.AuthReducer);

 // ----------------------- Validating the user -----------------------
 const [isUser, setIsUser] = useState(false);
 useEffect(() => {
  if (Number(problem.id_user) === Number(id)) {
   setIsUser(true);
  }
 }, [isUser, idUser, problem]);

 // ----------------------- Modal state hook -----------------------
 const [isOpenStatus, setIsOpenStatus] = useModal(false);

 //  ---------------- Handle apply ----------------
 let dataApplyProblem = {
  id_employee: idEmployee,
  id_employers_job_offer: idJobOffer,
  hired: 0,
 };
 const [isOpenApply, setIsOpenApply] = useModal(false);
 const [price, setPrice] = useInputForm(0);
 useEffect(() => {
  dataApplyProblem = {
   ...dataApplyProblem,
   price: price,
  };
  setDataApply(dataApplyProblem);
 }, [price]);

 // ---------------- Handle on qualificate to the employe ----------------
 const [jobOffer, setJobOffer] = useState({});
 const [fullJob, setFullJob] = useState({});
 useEffect(() => {
  const getOfferInformation = async () => {
   try {
    const fullProblem = await getFullProblem(problem.id);
    setFullJob(fullProblem.body);
    const data = await getJobOffer(problem.id_employer_job_offer);
    setJobOffer({ ...data.body });
   } catch (err) {
    console.error(err);
   }
  };
  if (idJobOffer || problem.id_employer_job_offer) {
   getOfferInformation();
  }
 }, [problem.id_employer_job_offer, idJobOffer]);
 const [dataQualificate, setDataQualificate] = useState({});
 const [isLoadingScore, setIsLoadingScore] = useState(false);
 const [isQualificated, setIsQualificated] = useState(false);
 const onQualificate = async () => {
  const dataScore = {
   ...dataQualificate,
   id_employer: idUser,
   id_employee: 15,
  };
  try {
   setIsLoadingScore(true);
   const resp = await postScoreUser(dataScore);
   setIsQualificated(true);
   setIsLoadingScore(false);
  } catch (err) {
   setIsLoadingScore(false);
   console.error(err);
  }
 };

 // ----------------------- Get workers of job offers -----------------------
 const [workersJob, setWorkersJob] = useState({});
 const { id_employer } = useSelector((state) => state.ProfileReducer);
 useEffect(() => {
  const getWorkers = async () => {
   try {
    const data = await getWorkersJob(id_employer);
    setWorkersJob(data.body);
   } catch (err) {}
  };
  if (id_employer) {
   getWorkers();
  }
 }, [id_employer]);

 // ---------------- defining UI of button action ----------------
 const defineAction = () => {
  if (problem.employer_job_offer_status.trim() === "available")
   return (
    <PostProblemStatus status={problem.employer_job_offer_status.trim()} />
   );
  if (!isUser)
   return (
    <Button active onClick={setIsOpenApply}>
     Apply!
    </Button>
   );
 };

 const RenderDetail = () => (
  <>
   <PostHeader
    name={problem.employer_name}
    avatar={problem.employer_avatar}
    profession={problem.employer_profession}
   />
   <PostProblemPrevisualization
    title={problem.title}
    idProblem={problem.id}
    short_description={problem.short_description}
    modality={problem.modality}
    location={problem.location}
   />
   <PostProblemDescription description={problem.long_description} />
   <PostProblemRequirements requirements={problem.requeriments} />
   <PostProblemSchedule schedule={problem.schedule} />

   {!isUser && problem.employer_job_offer_status ? (
    <>
     <PostProblemPayment
      salaryMax={problem.salary_range2}
      salaryMin={problem.salary_range1}
      setState={setPrice}
      state={price}
     />
    </>
   ) : (
    <PostProblemPayment agreed agreedPrice={problem.salary_range2} />
   )}

   <div className="post-problem-detail__actions">{defineAction()}</div>
   {isUser && problem.employer_job_offer_status.trim() === "available" && (
    <Button active onClick={setIsOpenStatus}>
     {" "}
     Status{" "}
    </Button>
   )}
  </>
 );

 // -------------------------------- RENDER OF THE COMPONENT --------------------------------
 return (
  <article className="post-problem-detail">
   {isError ? (
    <span className="post-problem-detail__error">{isError}</span>
   ) : isLoading ? (
    <Loading />
   ) : (
    RenderDetail()
   )}

   {/* // ---------------- Modal Status problem */}
   {isOpenStatus && (
    <ModalContainer>
     <Modal title="Status Of Problem" onClose={setIsOpenStatus}>
      <ProblemStatusInformation
       isUser={isUser}
       picture={problem.user_avatar}
       problemEmployer={problem.employer_name}
       problemWorker={"@"}
       setInformation={setDataQualificate}
       onQualificate={onQualificate}
       workersApplied={workersJob}
      />
     </Modal>
    </ModalContainer>
   )}

   {/* // ---------------- Modal Apply problem */}
   {isOpenApply && (
    <ModalContainer>
     <Modal title="Apply To Problem!" onClose={setIsOpenApply}>
      <CreateApplyPostProblem
       price={price}
       schedule={problem.schedule}
       onCancel={setIsOpenApply}
       onApply={onApply}
      />
     </Modal>
    </ModalContainer>
   )}

   {/* // ---------------- Modal loading score */}
   {isLoadingScore && (
    <ModalContainer>
     <Modal title="Scoring the solution!" onClose={setIsOpenApply}>
      <>
       <Loading />
       <span>Wait a moment while we validate your data</span>
      </>
     </Modal>
    </ModalContainer>
   )}
   {/* // ---------------- Modal score problem and worker */}
   {isQualificated && (
    <ModalContainer>
     <Modal title="Scoring the solution!" onClose={setIsOpenApply}>
      <ModalMessage
       type="great"
       message="Awesome! The problem was solved and scored!"
      />
     </Modal>
    </ModalContainer>
   )}
  </article>
 );
};

export default PostProblemDetail;
