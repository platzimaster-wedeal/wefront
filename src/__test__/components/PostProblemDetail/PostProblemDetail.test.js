import React from "react";
import { mount } from "enzyme";
import { create } from "react-test-renderer";
import PostProblemDetail from "../../../components/PostProblemDetail/PostProblemDetail";

describe("Test for the component PostProblemDetail :D", () => {
 const PostProblemDetailMounted = mount(<PostProblemDetail />);
 const PostProblemDetailCreated = create(<PostProblemDetail />);

 describe("Visual tests for the component", () => {
  test("should the comopnent have the same tree", () => {
   expect(PostProblemDetailCreated.toJSON()).toMatchSnapshot();
  });
 });

 describe("Functional tests for the component", () => {
  test("Should the component render one element", () => {
   expect(PostProblemDetailMounted.length).toEqual(1);
  });
  test("Should the content of the action string", () => {
   expect(
    typeof mount(<PostProblemDetail isWorker />)
     .find(".post-problem-detail__actions button")
     .text()
   ).toEqual("string");
  });
  test("Should the content of the action be Apply!", () => {
   expect(
    mount(<PostProblemDetail isWorker />)
     .find(".post-problem-detail__actions button")
     .text()
   ).toEqual("Apply!");
  });
 });
});
