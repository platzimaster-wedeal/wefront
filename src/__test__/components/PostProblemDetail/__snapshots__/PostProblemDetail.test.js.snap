// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the component PostProblemDetail :D Visual tests for the component should the comopnent have the same tree 1`] = `
<article
  className="post-problem-detail"
>
  <section
    className="post-header"
  >
    <figure
      className="post-header__picture"
      onClick={[Function]}
    >
      <img
        className="image image__isOnline undefined"
        src={Object {}}
      />
    </figure>
    <div
      className="post-header__user-info"
    >
      <div
        className="post-header__user-info--main"
      >
        <h3
          onClick={[Function]}
        >
          Alan Walks
        </h3>
        <span>
          Has a new problem!
        </span>
      </div>
      <strong
        className="post-header__user-info--profession"
      >
        <span
          className="label label__inactive"
        >
          Engineer
        </span>
      </strong>
    </div>
    <small
      className="post-header__time"
    >
      1hr
    </small>
  </section>
  <div
    className="post-problem-previsualization"
  >
    <div
      className="post-problem-previsualization__info"
    >
      <h4>
        Error with my pc
      </h4>
      <strong>
        #121515
      </strong>
    </div>
    <p
      className="post-problem-previsualization__description"
    >
      I need some help with my problem
    </p>
    <div
      className="post-problem-graphic-data"
    >
      <div
        className="graphic-data graphic-data--active"
      >
        <svg
          fill="currentColor"
          height={30}
          size={30}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width={30}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C8.13 2 5 5.13 5 9c0 5.25 7 13 7 13s7-7.75 7-13c0-3.87-3.13-7-7-7zm0 9.5c-1.38 0-2.5-1.12-2.5-2.5s1.12-2.5 2.5-2.5 2.5 1.12 2.5 2.5-1.12 2.5-2.5 2.5z"
          />
        </svg>
        <strong>
          COL
        </strong>
      </div>
      <div
        className="graphic-data undefined"
      >
        <svg
          fill="currentColor"
          height={30}
          size={30}
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width={30}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 11c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 2c0-3.31-2.69-6-6-6s-6 2.69-6 6c0 2.22 1.21 4.15 3 5.19l1-1.74c-1.19-.7-2-1.97-2-3.45 0-2.21 1.79-4 4-4s4 1.79 4 4c0 1.48-.81 2.75-2 3.45l1 1.74c1.79-1.04 3-2.97 3-5.19zM12 3C6.48 3 2 7.48 2 13c0 3.7 2.01 6.92 4.99 8.65l1-1.73C5.61 18.53 4 15.96 4 13c0-4.42 3.58-8 8-8s8 3.58 8 8c0 2.96-1.61 5.53-4 6.92l1 1.73c2.99-1.73 5-4.95 5-8.65 0-5.52-4.48-10-10-10z"
          />
        </svg>
        <strong>
          Remote
        </strong>
      </div>
    </div>
  </div>
  <section
    className="post-problem-description"
  >
    <strong
      className="post-problem-description__title"
    >
      Description
    </strong>
    <p
      className="post-problem-description__content"
    >
      Here must be some long description of the problem, thank you, next!
    </p>
  </section>
  <section
    className="post-problem-requirements"
  >
    <strong
      className="post-problem-requirements__title"
    >
      Minimun Requirements
    </strong>
    <ul
      className="post-problem-requirements__list"
    >
      <li>
        <div
          className="requirements__container"
        >
          <svg
            className="requirements__container-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
            />
          </svg>
          <p
            className="requirements__container-content"
          >
            Knowledge about PCs
          </p>
        </div>
      </li>
      <li>
        <div
          className="requirements__container"
        >
          <svg
            className="requirements__container-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 16.2L4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
            />
          </svg>
          <p
            className="requirements__container-content"
          >
            Knowledge about PCs
          </p>
        </div>
      </li>
    </ul>
  </section>
  <div
    className="post-problem-payment"
  >
    <label
      className="post-problem-payment__title"
      htmlFor="problem-price"
    >
      Payment
    </label>
    <p
      className="post-problem-payment__action-price"
    >
      Select your price:
       
      <span
        className="post-problem-payment__current-price"
      >
        $
        0
      </span>
    </p>
    <div
      className="post-problem-payment__range-price"
    >
      <small>
        min: 
        <span>
           $0 
        </span>
      </small>
      <small>
        max: 
        <span>
           $1000 
        </span>
      </small>
    </div>
    <input
      max={1000}
      min={0}
      onChange={[Function]}
      type="range"
    />
  </div>
  <div
    className="post-problem-detail__actions"
  />
  <button
    className="button button__primary undefined"
  >
    Edit!
  </button>
</article>
`;
